### DISCUSSION TITLE
File Library Organization

### DISCUSSION THREAD

**Discussion by uzziellite** [directus/directus#8950] [Feature Requests]
### The main problem
I believe that the current organization for files uploaded via Directus is not good. All the files stored on disk are bundled within one directory. In a scenario where someone is managing millions of files like pictures or other files, it will be quite difficult to organize this folder. Also, should someone do an accidental operation on the folder, all the files will be lost. 

### Solution
My suggestion is that the files should be stored in one parent directory with multiple sub directories each containing the files uploaded on monthly basis or weekly basis. At least that way file management from outside Directus is made easier.

**Comment by joselcvarela** [comment]
Hello @uzziellite 
I believe the uploads folder could be improved.
But how your suggestion could improve the accidental operations? I think the problem will persist.
If you want to avoid lose data, my recommendation is to use a service for backups like RAID for example.

Another idea is, instead of doing it by a monthly basis, we could project folders in File Library to filesystem.

**Reply by rijkvanzanten** [reply]
> we could project folders in File Library to filesystem.

I would personally prefer this over an arbitrary system like dates 👍🏻

**Reply by uzziellite** [reply]
Exactly, that's a much better way to do it. I would be glad to see this.

**Reply by stormbyte** [reply]
To lessen the strain on the application I would suggest using the file as indicator of the folder to be used.
Project folders would only seperate a single project.
If a project would have a lot of files as stated in #9701 the files per directory limit would hurt the performance.
Seperating the file in directories would prevent that.

My solution would be using the first and second letter of the file as subdirectories.
should result in `16^16` possible subdirectories.

As I am using it currently in a single project docker setup, I don't have multi project assets.

With multi project setup in mind having a `FileBackend` per project would be nice.
- ProjectA on digitalocean
- ProjectB on amazon
- ProjectC on local

**Reply by rijkvanzanten** [reply]
Interesting! It seems like there's a bit of a divide between people who'd prefer their file system to be "clean" (eg with mirrored folders as shown in Directus + using original file names), and people (like @stormbyte and myself) that treat the file storage as more of a blackbox that can be optimized for performance and "technical usage" rather than "friendly usage" 🤔 

Also regarding the file naming system for performance reasons, I think that would only matter for the `local` storage driver, as S3 (and therefore DigitalOcean Spaces) don't have "physical" folders at all, so there shouldn't be a limit of files per folder 🤔

**Reply by domdus** [reply]
I love the idea of having the file library folders -> reflected on the physical storage file system, well.. that would cause tears in my eyes. 😀 

I have about 385.000 files imported into one project, the generated thumbs or cropped images not included. I got things sorted in the file lib (8000 folders and 4 levels deep hierarchy) but the physical upload folder turned into a challenge.
Actually I avoid looking into it as FTP clients/ console freeze or crash when listing the files of that folder.

Not sure why I never imagined this as feature to ask for, but this would be superb!

**Reply by stormbyte** [reply]
I believe all file systems have limits, found this post on Stackoverflow
https://stackoverflow.com/questions/466521/how-many-files-can-i-put-in-a-directory

The virtual providers will have some magic solutions to stretch these limits but they are always present.

When a php project has a lot of session files reading the sessiondata takes a toll from the File System.
Seen it happen with 8Gb of session files. being forced to use `find` to list all the file and delete the oldest set is a pain.

I don't know how node handels these large directories. 
things like this: https://github.com/nodejs/node/commit/cbd8d715b2286e5726e6988921f5c870cbf74127
Might help handeling large directories.

Maybe in a full test create 100 million images of max 1Kb and see how this runs in this situation size does not matter.

**Reply by Pachat** [reply]
With thousands of projects, each one having a standardized identifier like 22-0001..., I'd prefer to be able to link the a path constructed on that identifier  22-0001 in folder 22/00/01, 22-0002  in folder 22/00/02,... 22-0941 in 22/09/41

**Comment by Dumpstar56** [comment]
I really appreciate this discussion! I would love to be able to access the folder structure of the file directory when selecting files via a file field in the app. I am not sure if there is also another way to make this happen, but I would sure appreciate that function!

**Reply by rijkvanzanten** [reply]
I think we could add a sidebar to the drawer-collection component in the app to dynamically render the same folder sidebar navigation you're used to seeing when selecting an item from directus_files 🤔 

Takes a bit of work, but that feels like the best approach! https://github.com/directus/directus/blob/c0e12638c651c9c704730693505938178a63ba32/app/src/views/private/components/drawer-collection.vue

**Reply by HemboPagi** [reply]
link does not seem to work

**Reply by rijkvanzanten** [reply]
@HemboPagi Updated the link above 👍🏻 FYI that was where this new feature would have to be implemented. I don't think anybody came long and actually did that yet

**Comment by HemboPagi** [comment]
Good discussion! I am currently working with the project where I need to ingest 28251 files (130GB) via Directus. Current data structure is built in a way where data item refers to the folder which has several images. So clean fliesystem would work nicely with that, but I guess I can come up some scripts to get it done also using the "blackbox" way .

**Reply by HemboPagi** [reply]
And now thinking more about it - we need to get images out one day as well as this is an archive and we need propably to hand over the conten after the project has finished. So what are te easiest ways to get data in and out? Only API scripts ?

**Reply by rijkvanzanten** [reply]
The fastest way to get that out would be to copy over the raw files stored on disk (or whatever storage provider you use), and then have a small script that renames those files back from the UUID directus uses to the originally uploaded file name. There's also bigger discussions going on on improving the underlying structure of the data store. Some relevant links:

https://github.com/directus/directus/discussions/12937

https://github.com/directus/directus/discussions/12980

https://github.com/directus/directus/discussions/14742#discussioncomment-3336717

**Reply by HemboPagi** [reply]
Thanks! I managed to copy all the data over with UUID based filenames and created also the entries to Directus files table. As the files are big, and some of them are in TIF format i assume that jpeg thumbnails need to be created as well. What would be the easy solution for that?

**Reply by rijkvanzanten** [reply]
> i assume that jpeg thumbnails need to be created as well

While using the Directus admin app, it'll auto generate those images on the fly if needed. There's no manual step required 👍🏻

**Reply by HemboPagi** [reply]
Maybe I have done something wrong but no thumbnails are created. I am browsing the Files Library and it loads full size images so it takes ages to load:
<img width="1073" alt="Screenshot 2022-10-18 at 09 13 16" src="https://user-images.githubusercontent.com/95413906/196349833-dacdaecf-cb77-4b61-94f1-96b6539517b4.png">
<img width="796" alt="Screenshot 2022-10-18 at 09 13 48" src="https://user-images.githubusercontent.com/95413906/196349940-df157640-b852-40d3-af57-99ef7c52fdb0.png">

**Reply by rijkvanzanten** [reply]
That's odd. It should download the system thumbnail presets for that view. Mind opening a new issue so we don't spam this unrelated discussion? Thanks! 🙂

**Reply by HemboPagi** [reply]
sure, opened  #16046

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
